Replace getPreviousPath with getPreviousRoute to get query params support

bogiii opened the pull request:
#### Changes proposed in this Pull Request

Support for taking query params from the previous path.

#### Testing instructions for dev environment

* Go to `/start/from/importing/wordpress?from={ORIGIN_WP_SITE}&to={SLUG}.wordpress.com`
  * ORIGIN_WP_SITE should have a Jetpack connection
  * {SLUG}.wordpress.com should be a simple site
* Select the "Import everything" option
* Press the "Start import" button
* It redirects to the "Checkout" screen
* From the checkout screen, click the close (X) button on the top-left corner
* Then, pick "Empty cart" or "Leave items" action
* Check if redirection to the previous page works (should be the same screen as it was before the checkout page)

Related to
- #59042
- #60241

bogiii committed:
Replace getPreviousPath with getPreviousRoute to get query params support

sirbrillig commented on client/my-sites/checkout/composite-checkout/lib/leave-checkout.ts:
```diff
@@ -1,6 +1,8 @@
 import { recordTracksEvent } from 'calypso/lib/analytics/tracks';
 import { navigate } from 'calypso/lib/navigate';
+import { addQueryArgs } from 'calypso/lib/route';
+import { QueryArgs } from 'calypso/my-sites/plans/jetpack-plans/types';
```
Minor: this can be `import type` since this is not used by any JS code.

```suggestion
import type { QueryArgs } from 'calypso/my-sites/plans/jetpack-plans/types';
```

sirbrillig commented on client/layout/masterbar/checkout.tsx:
```diff
@@ -9,12 +9,14 @@ import CalypsoShoppingCartProvider from 'calypso/my-sites/checkout/calypso-shopp
 import useValidCheckoutBackUrl from 'calypso/my-sites/checkout/composite-checkout/hooks/use-valid-checkout-back-url';
 import { leaveCheckout } from 'calypso/my-sites/checkout/composite-checkout/lib/leave-checkout';
 import useCartKey from 'calypso/my-sites/checkout/use-cart-key';
+import { QueryArgs } from 'calypso/my-sites/plans/jetpack-plans/types';
```
Same as below, this import can be stripped by the compiler:

```suggestion
import type { QueryArgs } from 'calypso/my-sites/plans/jetpack-plans/types';
```

sirbrillig reviewed (approved):
TIL about getPreviousQuery! I haven't tested this (I'll leave that to your team) but from a payments perspective this looks good to me.

vishnugopal reviewed (approved):
Ack! Confirming that redirection works, and works back to the page after clicking Start Import!

arthur791004 commented:
I have concern that is there any reason we don't use `previousRoute` (we can get this by `getPreviousRoute`) before? Adding this might make all of the checkout places support query params but I'm not sure is there any edge case to cause the problem 🤔

In addition, I think we should also add to the following places to make removing all of items from cart  be able to back to correct URL with query params
* https://github.com/Automattic/wp-calypso/blob/4309cd9b020a3c96c9cd896c6b7dad414f2ec2f3/client/my-sites/checkout/composite-checkout/composite-checkout.tsx#L691
* https://github.com/Automattic/wp-calypso/blob/4309cd9b020a3c96c9cd896c6b7dad414f2ec2f3/client/my-sites/checkout/composite-checkout/hooks/use-remove-from-cart-and-redirect.ts#L26

bogiii commented:
> I have concern that is there any reason we don't use previousRoute (we can get this by getPreviousRoute) before?

Thank you @arthur791004, it was much easier to consume `getPreviousSelector` (didn't know that exists) which contains the whole route with the query params.

I've rebased the changes and also updated two proposed paths with the full route selector.

arthur791004 commented:
https://github.com/Automattic/wp-calypso/pull/60476 is also related as we don't preserve query string in PayPal cancel_url

arthur791004 reviewed (approved):
Nice job 👍 The only concern is that some variables in the query string of the previous path might cause a problem, so we don't preserve the query string before. But I hope it's just my over consideration.
